crates/server/src/store/blocklist/mod.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// https://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// https://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+#![cfg(feature = "blocklist")]
+
+//! Blocklist resolver related types
+mod authority;
+
+pub use self::authority::BlocklistAuthority;
+use serde::Deserialize;
+
+/// Consult action enum.  Controls how consult lookups are handled.
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq)]
+pub enum BlocklistConsultAction {
+    /// Do not log or block any request when the blocklist is called via consult
+    Disabled,
+    /// Log and block matching requests when the blocklist is called via consult
+    Enforce,
+    /// Log but do not block matching requests when the blocklist is called via consult
+    Log,
+}
+
+/// Configuration for file based zones
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
+#[serde(default, deny_unknown_fields)]
+pub struct BlocklistConfig {
+    /// Support wildcards?  Defaults to true. If set to true, block list entries containing
+    /// asterisks will be expanded to match queries.
+    pub wildcard_match: bool,
+
+    /// Minimum wildcard depth.  Defaults to 2.  Any wildcard entries without at least this many
+    /// static elements will not be expanded (e.g., *.com has a depth of 1; *.example.com has a
+    /// depth of two.) This is meant as a safeguard against an errant block list entry, such as *
+    /// or *.com that might block many more hosts than intended.
+    pub min_wildcard_depth: u8,
+
+    /// Block lists to load.  These should be specified as relative (to the server zone directory)
+    /// paths in the config file.
+    pub lists: Vec<String>,
+
+    /// IPv4 sinkhole IP. This is the IP that is returned when a blocklist entry is matched for an
+    /// A query.  It defaults to 0.0.0.0.
+    pub sinkhole_ipv4: String,
```

Comment:
Suggest storing these as `Ipv4Addr`/`Ipv6Addr`?

One question is about where to default these. It feels potentially more idiomatic to make these `Option`, and then defer the default into the implementation (as long as there is no other dependency on these fields).

Reply:
The issue with this is that the error messages for an invalid address from serde are incomprehensible.  Storing them as a string and parsing/generating an error separately lets me have a message like 'you configured the address '257.0.0.1' that isn't a valid IPv4 address', whereas serde generates this gem:

```
1727879524:INFO:hickory_dns:443:loading configuration from: "tests/test-data/test_configs/chained_blocklist.toml"
Error: failed to read config file from "tests/test-data/test_configs/chained_blocklist.toml": toml decode error: TOML parse error at line 38, column 1
   |
38 | [[zones.stores]]
   | ^^^^^^^^^^^^^^^^
data did not match any variant of untagged enum StoreConfigContainer
```

Reply:
Hmm -- I think that has something to do with the container intermediate shenanigans, because it works fine in a simpler example:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2e742a3cd60ce35ad4b98add2224e041

```
Error {
    inner: Error {
        inner: TomlError {
            message: "invalid IP address syntax",
            raw: Some("ip = \"blah\""),
            keys: ["ip"],
            span: Some(5..11)
         }
    }
}
```

Reply:
This is indeed related to the way serde handles untagged enums. I'm going to change the blocklist to use Ipv4/6Addrs and put in a separate request to deal with the serde error messages.